experimental/list_formatter/src/deduplicating_array.rs

```diff
@@ -0,0 +1,111 @@
+use core::{mem, ops::Index, ptr};
+use icu_provider::yoke::*;
+
+#[derive(Debug)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Deserialize, serde::Serialize)
+)]
+// TODO: Make this generic over the 12. Serde currently can't derive
+// if const generics are present: https://github.com/serde-rs/serde/issues/1937
+pub struct DeduplicatingArray<T>([Dedupe<T>; 12]);
+
+#[derive(Debug, PartialEq)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Deserialize, serde::Serialize)
```

Comment:
Thought: Another way to achieve the deduplication goal would be to handle it all at the Serde level.

Keeping in mind that Postcard is not self-describing, you could write out the following Postcard constructions to deduplicate a list of `N` elements:

1. Number of unique elements `M`
2. Sequence of `N` elements from 0 to `M`
3. Sequence of `M` elements containing the actual data

When deserializing, use the following algorithm to make it zero-copy:

1. Read `M`
2. Read the sequence of indexes into a stack array of fixed length
3. Create an uninitialized stack array of elements
4. When reading each element from Serde, set it into each element of the list having that index according to the sequence of indexes you stored on the stack in step 2

Reply:
That sounds like a good solution. Currently I'm using untagged enums, so Serde tries to deserialize the variants from top to bottom and falls through if one fails. This means DeduplicatingArray<numeric> will probably not work, because it would deserialize the fallbacks as values (but that type doesn't make much sense anyway).

Reply:
Also, any overhead we add will be relatively expensive, because most patterns are < 10 bytes.